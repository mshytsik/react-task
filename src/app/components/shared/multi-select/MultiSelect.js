import React from "react";
import {
  FormControl,
  OutlinedInput,
  InputLabel,
  MenuItem,
  ListItemText,
  Select,
  Checkbox,
} from "@mui/material";

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 250,
    },
  },
};

const MultiSelect = ({ id, label, allValues, selectedValues, callback }) => {
  const handleChange = (event) => {
    const value = event.target.value;
    callback(typeof value === "string" ? value.split(",") : value);
  };

  return (
    <FormControl size="small" fullWidth className="multi-select">
      <InputLabel id={id + "-label"}>{label}</InputLabel>
      <Select
        labelId={id + "-label"}
        id={id}
        multiple
        value={selectedValues}
        onChange={handleChange}
        input={<OutlinedInput label={label} />}
        renderValue={(selected) => selected.join(", ")}
        MenuProps={MenuProps}
      >
        {allValues.map((value) => (
          <MenuItem key={value} value={value}>
            <Checkbox checked={selectedValues.indexOf(value) > -1} />
            <ListItemText primary={value} />
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

export default MultiSelect;
