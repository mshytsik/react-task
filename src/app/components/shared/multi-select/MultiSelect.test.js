import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

import MultiSelect from "./MultiSelect";

describe("test MultiSelect component", () => {
  const multiselect = (
    <MultiSelect
      id="multiselect"
      label="Multiselect Label"
      allValues={["A", "B", "C", "D"]}
      selectedValues={["C", "D"]}
      callback={() => {}}
    />
  );

  test("component is rendered correctly", () => {
    const component = renderer.create(multiselect).toJSON();
    expect(component).toMatchSnapshot();
  });

  test("check label text", () => {
    const { getByLabelText } = render(multiselect);
    const component = getByLabelText("Multiselect Label");
    expect(component).toBeInTheDocument();
  });

  test("selected values are checked", () => {
    const { container } = render(multiselect);
    const inputValues = container.querySelector("input").value;
    expect(["C", "D"].every((value) => inputValues.includes(value))).toBe(true);
  });
});
