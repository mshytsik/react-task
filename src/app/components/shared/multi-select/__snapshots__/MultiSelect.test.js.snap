// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MultiSelect component component is rendered correctly 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth multi-select css-q8hpuo-MuiFormControl-root"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    id="multiselect-label"
  >
    Multiselect Label
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="multiselect-label multiselect"
      className="MuiSelect-select MuiSelect-outlined MuiSelect-multiple MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      id="multiselect"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      C, D
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      disabled={false}
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value="C,D"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          Multiselect Label
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
